### ISSUE TITLE
Search engine optimization

### ISSUE THREAD

**Issue by tobyzerner** [flarum/core#189]
Flarum renders very basic versions of the discussion list and discussions (within `<noscript>` tags), which are going to be specifically optimised for search engines. This all takes place in Flarum\Forum\IndexAction and DiscussionAction, which extend Support\ClientAction and use Flarum\Support\ClientView. The actions get data from the Flarum's API and pass it onto content views (discussion.blade.php and index.blade.php). I'm happy to explain more if need be because it can get pretty confusing.

Put simply, the code is a mess. It needs to be rethought and refactored. A few thoughts:
- We probably want to use or build some kind of library that makes it easier to work with a JSON-API document payload.
- Similarly, we probably want to use or build some kind of library that will help generate URLs/query strings/etc.

Only after doing this will we be able to effectively optimise the SEO content and do the following:
- Add <link rel="canonical"> tags
- Add <meta> description tags
- Think about other stuff like schema.org
- https://moz.com/blog/seo-cheat-sheet

**Comment by axisdos** [comment]
Noindex for subpages (Inspired by Yoast SEO for WordPress) to avoid duplicate content. and URL Friendly.

**Comment by franzliedke** [comment]
@axisdos Noindex isn't needed when we have the `rel=canonical` tag. :)

**Comment by DCSJAPAN** [comment]
Questions of how to handle foreign language/unicode URLs -- such as https://github.com/flarum/core/issues/194 -- will also have SEO implications.

In that regard, unfortunately, I think different regions may prefer different solutions.

**Comment by neandrake** [comment]
Would this tie into Accessibility at all? I'm not extremely familiar with it, but when structuring the noscript HTML along with the rest of the page, it would be nice if parts of it were identifiable for users with disabilities.

http://www.w3.org/WAI/intro/aria

**Comment by franzliedke** [comment]
Could [this approach](http://www.sitepoint.com/cutting-the-mustard-with-css-media-queries/) be an alternative to the <noscript> tags?

**Comment by tobscure** [comment]
Quite possibly, feels a *little* bit hacky though? I'm not too fussed

**Comment by dcsjapan** [comment]
https://github.com/flarum/core/issues/906 is an SEO-related issue specific to the Tags extension.

**Comment by 01010000101001100** [comment]
I investigated links on my Flarum and posted results in this topic on official Flarum forum, have a look:
https://discuss.flarum.org/d/3103-is-the-flarum-s-seo-seriously-messed-up-or-am-i-doing-something-wrong

**Comment by franzliedke** [comment]
For reference: https://webmasters.googleblog.com/2014/02/infinite-scroll-search-friendly.html

**Comment by franzliedke** [comment]
Also interesting, with regards to changing `<title>` and other meta tags when doing client-side navigation: http://searchengineland.com/tested-googlebot-crawls-javascript-heres-learned-220157

**Comment by gwillem** [comment]
@01010000101001100 refers to a bug in the <noscript> pagination code. Version 0.1.0-beta.7 will infinitely link to the "next" page id, even if there are no more discussions. I have only 1 page of discussions, but Google has already crawled to page 584. Fixing this would probably improve SEO a lot.

(--> flarum/framework#1310)

**Comment by franzliedke** [comment]
@gwillem @01010000101001100 Thanks for the heads-up. I've [fixed](https://github.com/flarum/core/commit/6913e8f0f8ea98ffa89f8537a714580c1e6cffbd) the incorrect generation of pagination links, along with some other minor pagination-related improvements.

**Comment by askvortsov1** [comment]
To clarify, what exactly needs to be done here? Based off of the link Franz sent, Google seems to be fully capable of handling our frontend. Is it just a question of ensuring pagination is properly done?

I've been resolving meta title and description issues through a series of PRs, see https://github.com/flarum/core/issues/495#issuecomment-594183279, https://github.com/flarum/core/pull/2049

**Comment by BartVB** [comment]
Google is able to render Javascript but this is much, much slower than indexing 'static' HTML. More information here:
https://ja.dev/entry/blog/nagayama/render-budget-en

TLDR; use Javascript for endusers, HTML for crawlers.

**Comment by franzliedke** [comment]
Also, Google isn't the only search engine, especially in some other countries.

**Comment by franzliedke** [comment]
You're right, though, that we need to spec out more clearly what aspects we want to tackle. "SEO" is a very generic and not very actionable issue title. :laughing:

**Comment by flarumalshain** [comment]
**There are too many dofollow links on the site. ("Mentions, Likes, Replies")** - Their attributes must be "rel = nofollow".
This item is very, very important. The reason for this: When Google bots enter the page, they see too many dofollow links and try to index them.

**the ?near tag.**
Google indexes both the near tag and the normal page. Therefore, because the two urls are trying to index in the same way, the bot is slow and cannot navigate the site in detail.

**Comment by askvortsov1** [comment]
Splitting into https://github.com/flarum/core/issues/2397, https://github.com/flarum/core/issues/626, https://github.com/flarum/core/issues/2397
